### ISSUE TITLE
MAUI very slow on physical android devices

### ISSUE THREAD

**Issue by OpterJohan-Hallberg** [dotnet/maui#18838]
### Description

MAUI Android runs very well on emulators, but on physical devices, both debug and release it runs very slow. I believe its due to an error that gets throw due to a variety of reasons, that gets worse the more content a project has. I'm our current project just switching to another page can take up to 10 or so seconds, on a new MAUI project it isn't noticeable, but its there still. This did not use to take that long i Xamarin. Its unusable with that much delay on pretty much everything.

The errors always look something similar to this:
`[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'androidx/fragment/app/FragmentManagerImpl'
[monodroid-assembly] typemap: called from
[monodroid-assembly] at Java.Interop.TypeManager.GetJavaToManagedType(String class_name) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 228
[monodroid-assembly]    at Java.Interop.TypeManager.CreateInstance(IntPtr handle, JniHandleOwnership transfer, Type targetType) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 274
[monodroid-assembly]    at Java.Lang.Object.GetObject(IntPtr handle, JniHandleOwnership transfer, Type type) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 303
[monodroid-assembly]    at Java.Lang.Object._GetObject[Frag`

It can be reproduced in a completely new and untouched MAUI project too:
`[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'com/android/internal/policy/PhoneLayoutInflater'
[monodroid-assembly] typemap: called from
[monodroid-assembly] at Java.Interop.TypeManager.GetJavaToManagedType(String class_name) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 230
[monodroid-assembly]    at Java.Interop.TypeManager.CreateInstance(IntPtr handle, JniHandleOwnership transfer, Type targetType) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 274
[monodroid-assembly]    at Java.Lang.Object.GetObject(IntPtr handle, JniHandleOwnership transfer, Type type) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 304
[monodroid-assembly]    at Java.Lang.Object._GetObject[LayoutInflater](IntPtr handle, JniHandleOwnership transfer) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 290
[monodroid-assembly]    at Java.Lang.Object.GetObject[LayoutInflater](IntPtr handle, JniHandleOwnership transfer) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 281
[monodroid-assembly]    at AndroidX.Fragment.App.Fragment.n_OnCreateView_Landroid_view_LayoutInflater_Landroid_view_ViewGroup_Landroid_os_Bundle_(IntPtr jnienv, IntPtr native__this, IntPtr native_inflater, IntPtr native_container, IntPtr native_savedInstanceState) in /Users/runner/work/1/s/generated/androidx.fragment.fragment/obj/Release/net6.0-android/generated/src/AndroidX.Fragment.App.Fragment.cs:line 2028
[monodroid-assembly]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPLLL_L(_JniMarshal_PPLLL_L callback, IntPtr jnienv, IntPtr klazz, IntPtr p0, IntPtr p1, IntPtr p2) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 352
[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'androidx/fragment/app/FragmentManagerImpl'
[monodroid-assembly] typemap: called from
[monodroid-assembly] at Java.Interop.TypeManager.GetJavaToManagedType(String class_name) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 230
[monodroid-assembly]    at Java.Interop.TypeManager.CreateInstance(IntPtr handle, JniHandleOwnership transfer, Type targetType) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 274
[monodroid-assembly]    at Java.Lang.Object.GetObject(IntPtr handle, JniHandleOwnership transfer, Type type) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 304
[monodroid-assembly]    at Java.Lang.Object._GetObject[FragmentManager](IntPtr handle, JniHandleOwnership transfer) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 290
[monodroid-assembly]    at Java.Lang.Object.GetObject[FragmentManager](IntPtr handle, JniHandleOwnership transfer) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 281
[monodroid-assembly]    at AndroidX.Fragment.App.Fragment.get_ChildFragmentManager() in /Users/runner/work/1/s/generated/androidx.fragment.fragment/obj/Release/net6.0-android/generated/src/AndroidX.Fragment.App.Fragment.cs:line 416
[monodroid-assembly]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellFragmentContainer.OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) in D:\a\_work\1\s\src\Controls\src\Core\Compatibility\Handlers\Shell\Android\ShellFragmentContainer.cs:line 48
[monodroid-assembly]    at AndroidX.Fragment.App.Fragment.n_OnCreateView_Landroid_view_LayoutInflater_Landroid_view_ViewGroup_Landroid_os_Bundle_(IntPtr jnienv, IntPtr native__this, IntPtr native_inflater, IntPtr native_container, IntPtr native_savedInstanceState) in /Users/runner/work/1/s/generated/androidx.fragment.fragment/obj/Release/net6.0-android/generated/src/AndroidX.Fragment.App.Fragment.cs:line 2031
[monodroid-assembly]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPLLL_L(_JniMarshal_PPLLL_L callback, IntPtr jnienv, IntPtr klazz, IntPtr p0, IntPtr p1, IntPtr p2) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 352
[Mono] Running class .cctor for Microsoft.Maui.Handlers.PageHandler from '/data/data/com.companyname.mauiapp3test/files/.__override__/Microsoft.Maui.dll'
[Mono] Running class .cctor for Microsoft.Maui.Handlers.ContentViewHandler from '/data/data/com.companyname.mauiapp3test/files/.__override__/Microsoft.Maui.dll'`

### Steps to Reproduce

1. Create a new .Net MAUI App
2. Start debug to physical device.
3. Give i a moment after it launches
4. `[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'com/android/internal/policy/PhoneLayoutInflater' error will appear in debug log.

### Version with bug

8.0.3

### Is this a regression from previous behavior?

Yes, this used to work in Xamarin.Forms

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 12 and up at least.

### Did you find any workaround?

We have not been able to find a workaround.

### Relevant log output

```shell
[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'com/android/internal/policy/PhoneLayoutInflater'
[monodroid-assembly] typemap: called from
[monodroid-assembly] at Java.Interop.TypeManager.GetJavaToManagedType(String class_name) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 230
[monodroid-assembly]    at Java.Interop.TypeManager.CreateInstance(IntPtr handle, JniHandleOwnership transfer, Type targetType) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 274
[monodroid-assembly]    at Java.Lang.Object.GetObject(IntPtr handle, JniHandleOwnership transfer, Type type) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 304
[monodroid-assembly]    at Java.Lang.Object._GetObject[LayoutInflater](IntPtr handle, JniHandleOwnership transfer) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 290
[monodroid-assembly]    at Java.Lang.Object.GetObject[LayoutInflater](IntPtr handle, JniHandleOwnership transfer) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 281
[monodroid-assembly]    at AndroidX.Fragment.App.Fragment.n_OnCreateView_Landroid_view_LayoutInflater_Landroid_view_ViewGroup_Landroid_os_Bundle_(IntPtr jnienv, IntPtr native__this, IntPtr native_inflater, IntPtr native_container, IntPtr native_savedInstanceState) in /Users/runner/work/1/s/generated/androidx.fragment.fragment/obj/Release/net6.0-android/generated/src/AndroidX.Fragment.App.Fragment.cs:line 2028
[monodroid-assembly]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPLLL_L(_JniMarshal_PPLLL_L callback, IntPtr jnienv, IntPtr klazz, IntPtr p0, IntPtr p1, IntPtr p2) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 352
[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'androidx/fragment/app/FragmentManagerImpl'
[monodroid-assembly] typemap: called from
[monodroid-assembly] at Java.Interop.TypeManager.GetJavaToManagedType(String class_name) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 230
[monodroid-assembly]    at Java.Interop.TypeManager.CreateInstance(IntPtr handle, JniHandleOwnership transfer, Type targetType) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 274
[monodroid-assembly]    at Java.Lang.Object.GetObject(IntPtr handle, JniHandleOwnership transfer, Type type) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 304
[monodroid-assembly]    at Java.Lang.Object._GetObject[FragmentManager](IntPtr handle, JniHandleOwnership transfer) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 290
[monodroid-assembly]    at Java.Lang.Object.GetObject[FragmentManager](IntPtr handle, JniHandleOwnership transfer) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 281
[monodroid-assembly]    at AndroidX.Fragment.App.Fragment.get_ChildFragmentManager() in /Users/runner/work/1/s/generated/androidx.fragment.fragment/obj/Release/net6.0-android/generated/src/AndroidX.Fragment.App.Fragment.cs:line 416
[monodroid-assembly]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellFragmentContainer.OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) in D:\a\_work\1\s\src\Controls\src\Core\Compatibility\Handlers\Shell\Android\ShellFragmentContainer.cs:line 48
[monodroid-assembly]    at AndroidX.Fragment.App.Fragment.n_OnCreateView_Landroid_view_LayoutInflater_Landroid_view_ViewGroup_Landroid_os_Bundle_(IntPtr jnienv, IntPtr native__this, IntPtr native_inflater, IntPtr native_container, IntPtr native_savedInstanceState) in /Users/runner/work/1/s/generated/androidx.fragment.fragment/obj/Release/net6.0-android/generated/src/AndroidX.Fragment.App.Fragment.cs:line 2031
[monodroid-assembly]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPLLL_L(_JniMarshal_PPLLL_L callback, IntPtr jnienv, IntPtr klazz, IntPtr p0, IntPtr p1, IntPtr p2) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 352
```

**Comment by asi-evin** [comment]
I can confirm I am getting this error all the time in debug. Common ones include (but not limited to):

- typemap: unable to find mapping to a managed type from Java type 'androidx/fragment/app/FragmentManagerImpl'
- typemap: unable to find mapping to a managed type from Java type 'android/app/SharedPreferencesImpl'
- typemap: unable to find mapping to a managed type from Java type 'android/app/ApplicationPackageManager'
- typemap: unable to find mapping to a managed type from Java type 'com/android/internal/policy/PhoneLayoutInflater'
- typemap: unable to find mapping to a managed type from Java type 'android/view/WindowManagerImpl'

That last one happens VERY often as it's triggered by getting the main display DPI in a looped animation (at Microsoft.Maui.Devices.DeviceDisplayImplementationBase.get_MainDisplayInfo()).

It makes it nearly impossible to find debug messages in the console.

**Comment by South2AK** [comment]
Having the same problem. Any news to this?

**Comment by pboisso** [comment]
Same here.

**Comment by rudyspano** [comment]
Forcing xaml compilation results in same performance as Xamarin Forms in our side in Debug. (Breaks Xaml Hot Reload for the moment)
https://www.sharpnado.com/net-maui-painfully-slow-debugging/?fbclid=IwAR0_aHJXi0OBoh8VO7gdn_u2AJcRBETKqgM05MnsaomaCmsFCnbq-GDGaBI

**Comment by clintcambier** [comment]
Also from our end there is a major difference in which physical Android devices are about **5 times slower** than Android emulators AND iOS physical devices.

This is a critical problem that does not seem to be on the radar at all. Who should we contact to escalate this further?

**Comment by VNGames** [comment]
@clintcambier @QiteSimon Did you try solution in the post above yours?

Adding this for DEBUG should solve slowness 
`<_MauiForceXamlCForDebug>true</_MauiForceXamlCForDebug>`
(I didn't try yet, Just asking but it seems like this should be the solution. In Production my app isn't slow at all. I would say even faster than Xamarin, But during DEBUG it is slow on a physical device Android and article above says it is because in MAUI xaml compilation is turned off by default)

**Comment by catnip050** [comment]
The same problem exists.
`<_MauiForceXamlCForDebug>true</_MauiForceXamlCForDebug>` did not solve the problem.
Has anyone made any progress?

**Comment by mnxamdev** [comment]
> I'm having trouble with this as well. My app, which was migrated from Xamarin, is MUCH slower in MAUI than in Xamarin. At least it is on Android. iOS seems to be running just fine.

Same here but to a lesser degree on iOS as it's definitely not as snappy as Xamarin.Forms on iOS. Our project on Xamarin.Forms in Android is much more performant yet Xamarin.Forms support is End of Life in 5 days. Menu drawer closing while navigating is supper laggy as well as everything else so much so that we can't even develop properly. Again, these issues are also apparent in Release mode and we can't release to production with these serious performance issues. They exist in our app in debug and release in MAUI but not Xamarin.Forms. Microsoft needs to extend Xamarin.Forms support until MAUI is stable and performant as the product they intend to be replacing.

**Comment by clintcambier** [comment]
Fully agree that support should be extended... Any update about that?

Op vr 26 apr. 2024 18:08 schreef mnxamdev ***@***.***>:

**Comment by rudyspano** [comment]
@mnxamdev, my advice: I know that can be time consuming but you should create small projects to isolate each problem that you suspect. I personnaly don't use "Menu Drawer" for instance.
If you can reproduce your performance issues in small  projects that you can share and create Github issue Microsoft and the community will help you...

**Comment by Zhanglirong-Winnie** [comment]
Verified this issue with Visual Studio 17.10 Preview 5 (8.0.21&8.0.3). Can repro on android device platform.
![image](https://github.com/dotnet/maui/assets/77434865/16acbc0d-ea01-4de5-a077-c89059292542)

**Comment by PureWeen** [comment]
Can you attach a logcat file?
https://learn.microsoft.com/en-us/xamarin/android/deploy-test/debugging/android-debug-log?tabs=windows

You can also grab a speedscope to see if that reveals anything
https://github.com/dotnet/maui/wiki/Profiling-.NET-MAUI-Apps#feature-xyz-was-faster-in-xamarinforms

**Comment by PureWeen** [comment]
@jonathanpeppers thoughts?

**Comment by SKeehnen** [comment]
A solution would be great, not only makes it the app slower but also produces an unreadable debug window. I'm using many settings through the Preferences, so my debug window is completely filled with these messages:

`[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'android/app/SharedPreferencesImpl'
[monodroid-assembly] typemap: called from
[monodroid-assembly] at Java.Interop.TypeManager.GetJavaToManagedType(String class_name) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 228
[monodroid-assembly]    at Java.Interop.TypeManager.CreateInstance(IntPtr handle, JniHandleOwnership transfer, Type targetType) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 274
[monodroid-assembly]    at Java.Lang.Object.GetObject(IntPtr handle, JniHandleOwnership transfer, Type type) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 303
[monodroid-assembly]    at Java.Lang.Object._GetObject[ISharedPreferences](IntPtr handle, JniHandleOwnership transfer) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 289
[monodroid-assembly]    at Java.Lang.Object.GetObject[ISharedPreferences](IntPtr handle, JniHandleOwnership transfer) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 281
[monodroid-assembly]    at Android.Preferences.PreferenceManager.GetDefaultSharedPreferences(Context context) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net8.0/android-34/mcw/Android.Preferences.PreferenceManager.cs:line 886
[monodroid-assembly]    at Microsoft.Maui.Storage.PreferencesImplementation.GetSharedPreferences(String sharedName) in D:\a\_work\1\s\src\Essentials\src\Preferences\Preferences.android.cs:line 160
[monodroid-assembly]    at Microsoft.Maui.Storage.PreferencesImplementation.Get[Int32](String key, Int32 defaultValue, String sharedName) in D:\a\_work\1\s\src\Essentials\src\Preferences\Preferences.android.cs:line 99
[monodroid-assembly]    at Microsoft.Maui.Storage.Preferences.Get(String key, Int32 defaultValue, String sharedName) in D:\a\_work\1\s\src\Essentials\src\Preferences\Preferences.shared.cs:line 185
[monodroid-assembly]    at Microsoft.Maui.Storage.Preferences.Get(String key, Int32 defaultValue) in D:\a\_work\1\s\src\Essentials\src\Preferences\Preferences.shared.cs:line 116`

**Comment by jonathanpeppers** [comment]
The `monodroid-assembly` log messages are not on by default. Did you run a command like this to enable them?

```
adb shell setprop debug.mono.log assembly
```

Simply enabling log messages  like these will make the app slower. `setprop` doesn't clear unless you reboot the device, but you can clear it with:

```
adb shell setprop debug.mono.log "''"
```

**Comment by SKeehnen** [comment]
Thanks @jonathanpeppers that worked perfect!

**Comment by wagaana** [comment]
My app was working well in .net 7.0 preview, but after upgrading it to .net 8.0, the app got slow on Android.

**Comment by jonathanpeppers** [comment]
@wagaana did you profile your app? If you could share a sample project or `.speedscope` file: https://aka.ms/profile-maui

I don't know of any performance regression going from .NET 7 to .NET 8.

**Comment by irshadka** [comment]
Any fixes for this…… iOS running perfectly but android full lag.

**Comment by marcmarsinach** [comment]
> I'm having trouble with this as well. My app, which was migrated from Xamarin, is MUCH slower in MAUI than in Xamarin. At least it is on Android. iOS seems to be running just fine.

Same problem here

**Comment by jonathanpeppers** [comment]
@marcmarsinach did you already review some of the instructions at the top here: https://aka.ms/profile-maui

**Comment by schaudhary-extron** [comment]
Is this issue resolved? I am still having the issue with application not finding mapping and slow debugging.

**Comment by jonathanpeppers** [comment]
> Any news here? Have same problem, in emulator is ok, but on physical android device unusable

This sounds backwards to me: my experience is that something like even a Pixel 5 (few years old) is faster than an emulator.

If someone is able to review https://aka.ms/profile-maui and share `.speedscope` files of the slowdown we can investigate. Thanks!

**Comment by dijonsmit** [comment]
Also experiencing the same issue, Android version of our app is extremely laggy and will give our customers very bad experience and not good for the companies reputation.

**Comment by marcmarsinach** [comment]
Solved replacing autommaper with mapster

**Comment by KevThompson2020** [comment]
Same issue here.  CollectionView click can take something like half a second if not more to register on Android ( Pixel 7 so not exactly a slow device ).  IOS, Windows, Mac are all fine.

**Comment by jonathanpeppers** [comment]
> Same issue here. CollectionView click can take something like half a second if not more to register on Android ( Pixel 7 so not exactly a slow device ). IOS, Windows, Mac are all fine.

I'm confused by this comment, as clicking should not do anything. It has 1 Java -> C# transition, and allocates a `Android.Graphics.Rect`. Can you share a `.speedscope` of this slowdown? https://aka.ms/profile-maui

**Comment by jlinkerlathem** [comment]
I also am experiencing this.  I have done a speed scope file of just a simple button click.  It loads a screen that has 12 buttons on it.  They are not actual buttons, but are custom controls.  It takes it about 4 or 5 seconds to load the buttons (removed xaml from the equation and load the buttons via code), then spends another 5 seconds in Microsoft.Maui.Controls!Microsoft.Maui.Controls.Element.SetHandler.

**Comment by jonathanpeppers** [comment]
@jlinkerlathem can you share the file?

**Comment by jonathanpeppers** [comment]
To be clear, we aren't saying button clicks are slow. We are saying navigation to new screens are slow?

In your app, there are a few things that you could improve:
* `CB.Logic!CB.Logic.ViewModels.BaseViewModel.UpdateNetworkConnectivityStatus()` looks expensive, does every one of your `ViewModel`s call it? Could that be a one time thing at startup?
* `CB.Maui.Common!CB.Maui.Common.Controls.LathemButton.InitializeComponent()` looks expensive, is the XAML complicated? Or are there a lot of them? The XAML does look compiled, are there any warnings related to `x:DataType` on the bindings?
* `MR.Gestures.dll` seems a bit expensive, in general. Can some of the built-in gestures be used instead?

There is a lot of interesting things in here, in general, though. Thanks!

**Comment by jlinkerlathem** [comment]
@jonathanpeppers 

Yes, it is the navigation I am referring to.

- UpdateNetworkConnectivityStatus() - I can disable this for testing.  We have icons at the top of each screen that indicate the current connectivity status.  All it does is use Microsoft.Maui.Networking.Connectivity.NetworkAcess to check if connected, and Microsoft.Maui.Networking.Connectivity.ConnectionProfiles to determine if WiFi, then sets an icon to ImageSource.FromFile based on the results.
- CB.Maui.Common!CB.Maui.Common.Controls.LathemButton.InitializeComponent() - This was what I honed in on.  I tried to remove xaml as the culprit so for the screen i was testing with I switched to a LathemSoundButton that is the same thing but instead of xaml it builds the UI in code (just a grid with the buttons in it).  Do you see any issues with the LathemSoundButton?  I haven't figured out how to search in SpeedScope, so I need to find where it is still using the LathemButton and switch them all to the Sound one.
- MR.Gestures.dll - We are actually in the process of removing this, I will go ahead and do this to see what affect it has

Glad you see interesting things.  If you see anything else useful let me know.  I will swap out the remaining LathemButtons for the LathemSound ones, disable the UpdateNetworkConnectivityStatus(), and remove Mr.Gestures and see what affect that has, if I still see the issue I will submit another speed scope file

**Comment by jonathanpeppers** [comment]
If you know what method to look for, use the Sandwich view, search box, and then the call stack should show who called it:

![Image](https://github.com/user-attachments/assets/5687b869-8517-457f-b7e2-282b1d14c40a)

**Comment by jlinkerlathem** [comment]
@jonathanpeppers Well, that would be why I couldn't figure out how to search, mine does not have a search icon there...

![Image](https://github.com/user-attachments/assets/18cb7638-5d0d-484e-a617-d99cb1f93574)

**Comment by jlinkerlathem** [comment]
@jonathanpeppers Ok, I removed Mr. Gestures.  Also removed the UpdateNetworkConnectivityStatus() call.  I also moved all of the custom controls involved out of XAML and to be created in C#.  I thought it had sped it up one of the times I ran it, but I must have done something wrong as I am still getting slow times.  I have attached the latest Speed Scope file.  It seems to me the LathemSoundButton is the issue.  But I cannot figure out why.  As mentioned, I am building it in code.   

Its a Border, with a RoundRectangle (with two data triggers), then a Grid (with a TapGestureRecognizer (handles the bound command), and a TouchBehavior (handles swapping the image source to a "pressed" state image if an image has been added) attached), and then an Image and a Label in the grid (each with some bindings).  I could see the button being too bulky being the the issue if the problem was worse with more buttons, but it seems even one button on the screen kills it.  Any more thoughts?

 [dotnet-dsrouter.exe_20241101_132904.speedscope.json](https://github.com/user-attachments/files/17602580/dotnet-dsrouter.exe_20241101_132904.speedscope.json)

**Comment by jonathanpeppers** [comment]
If `LathemSoundButton` is created completely in code, maybe you can get rid of all `Binding` and `DataTrigger`?

I have an example here, but maybe the control can override `OnBindingContextChanged()` and just set the values without bindings?

* https://github.com/jonathanpeppers/Simulations/blob/d5c068136ebc1a90b52f5a9523be86d09f73c5f1/2D/Simulations.Mobile/OrganismView.cs#L16

In this example, a new `Organism` (imagine it's a model or viewmodel) would come in, set some properties. For changes within a single `Organism`, I made the `Organism` fire a single `PropertyChanged` with an empty string to update. I don't know what your button does, but I bet this would help quite a bit.

Reviewing your logs, I also see how fixing this would help:

* https://github.com/dotnet/android/issues/7479

So, I'm looking into it.

**Comment by jlinkerlathem** [comment]
@jonathanpeppers Hmmm, that could work with this particular control as I do not think the values change after initialization.  I guess what confuses me though is what is so different between Maui's handling of this (bindings, xaml creation/parsing, whatever else is involved here) that would make it take so vastly different an amount of time as it did from Xamarin?

I will try the without binding option, but I am worried it will have a negligible affect.  Will let you know what I find.  I will be starting on testing a release of our software on Monday, so if I have not responded by then it may be a few days before I can circle back.

**Comment by jlinkerlathem** [comment]
@jonathanpeppers Also, before I forget, I want to express my genuine appreciation for your persistent assistance in tracking this down

**Comment by jlinkerlathem** [comment]
@jonathanpeppers I don't think it is the bindings.  Just to see I removed all of the bindings, DataTriggers, Behaviors, the only thing I left was the TapGestureRecognizer to be able to tap on a button and navigate, and the Text Property.  I removed EVERYTHING else, and it still takes the same amount of time...  I am running out of ideas as everything I have tried has had no affect on it.  Like, not even a little.  I would at least expect these changes to be speeding it up a little here and a little there.  It takes it about 10 seconds from the time I tap the button until the next page loads, and that time has remained consistent through every change I have made....

**Comment by jlinkerlathem** [comment]
I switched LathemSoundButton to be just a normal button instead of a content view.  Loses my functionality but for testing purposes helps narrow something down.  It did speed it up a little.  UI creation went from about 3 or so seconds down to 2.25, but there is still a 4 second section of it doing the LayoutHandlers.  I will say with the Button instead of Content View, it does load fast the second time you call the page (about 6.5 seconds as opposed to 8, where as with the Content View button it did not speed up after the first time).  I think I have stared at this too long now though...

**Comment by jonathanpeppers** [comment]
@jlinkerlathem how many controls are on the page you are testing? Is it in the hundreds? (The joke is you should have less views than pixels on the screen)

Also keep in mind the numbers you see in `dotnet-trace` / .speedscope can be as much as 2x larger than real life with all the instrumentation turned off.

**Comment by jlinkerlathem** [comment]
@jonathanpeppers  Visually there are less than 20 items on the screen, 12 buttons, a text box, a title bar with a label and a back button, and then a status bar at the top with the time.  There may be a few more objects as containers for those objects, but its not a list or anything like that.  Just a grid.

I have ran it in release mode to verify the instrumentation was not the culprit and it still takes the same amount of time to laod the page.

As I mentioned, for the next few days I will be on release testing another app, once that is complete I am going to flatten the UI for this page and see what that does.  Currently its a ContentPage that is loaded into a BaseContentPage (has the title bar and status bar that is shared by all of the pages), then it loads a Custom Control (a NumberPad) which has 12 of the LathemSoundButtons on it.  I am going to flatten that so there is no base page, just all of the stuff on one page, and the number pad code will reside in the parent page and the buttons will be copied on the page and see if that speeds it up.  If it does then I will separate things back out one at a time until I hit the issue, and go from there.  I will try to update when I get to that point to help others, or if it doesn't solve it when I flatten it I will post another speed scope file.

**Comment by KevThompson2020** [comment]
> > > Same issue here. CollectionView click can take something like half a second if not more to register on Android ( Pixel 7 so not exactly a slow device ). IOS, Windows, Mac are all fine.
> > 
> > 
> > I'm confused by this comment, as clicking should not do anything. It has 1 Java -> C# transition, and allocates a `Android.Graphics.Rect`. Can you share a `.speedscope` of this slowdown? https://aka.ms/profile-maui
> 
> I'll run a speed scan when I get change ( probably later this week due to a deadline ). when running in release mode it's a lot better but still feels a little sluggish. in debug mode it takes a good chunk of time to register the click, even to the point of showing the android ripple effect.

I have tried running a speed test this afternoon and its having some issues running.

when I run dotnet-trace collect -p 21001 --format speedscope it crashes.

No profile or providers specified, defaulting to trace profile 'cpu-sampling'

Provider Name                           Keywords            Level               Enabled By
Microsoft-DotNETCore-SampleProfiler     0x0000F00000000000  Informational(4)    --profile 
Microsoft-Windows-DotNETRuntime         0x00000014C14FCCBD  Informational(4)    --profile 

[ERROR] System.IO.EndOfStreamException: Unable to read beyond the end of the stream.
   at System.IO.Stream.ReadAtLeastCore(Span`1 buffer, Int32 minimumBytes, Boolean throwOnEndOfStream)
   at System.IO.BinaryReader.InternalRead(Int32 numBytes)
   at System.IO.BinaryReader.ReadUInt16()
   at Microsoft.Diagnostics.NETCore.Client.IpcHeader.Parse(BinaryReader reader) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcHeader.cs:line 55
   at Microsoft.Diagnostics.NETCore.Client.IpcMessage.Parse(Stream stream) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcMessage.cs:line 117
   at Microsoft.Diagnostics.NETCore.Client.IpcClient.Read(Stream stream) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcClient.cs:line 107
   at Microsoft.Diagnostics.NETCore.Client.IpcClient.SendMessageGetContinuation(IpcEndpoint endpoint, IpcMessage message) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcClient.cs:line 44
   at Microsoft.Diagnostics.NETCore.Client.EventPipeSession.Start(IpcEndpoint endpoint, EventPipeSessionConfiguration config) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/EventPipeSession.cs:line 45
   at Microsoft.Diagnostics.NETCore.Client.DiagnosticsClient.StartEventPipeSession(EventPipeSessionConfiguration config) in /_/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsClient/DiagnosticsClient.cs:line 101
   at Microsoft.Diagnostics.Tools.Trace.CollectCommandHandler.Collect(CancellationToken ct, IConsole console, Int32 processId, FileInfo output, UInt32 buffersize, String providers, String profile, TraceFileFormat format, TimeSpan duration, String clrevents, String clreventlevel, String name, String diagnosticPort, Boolean showchildio, Boolean resumeRuntime, String stoppingEventProviderName, String stoppingEventEventName, String stoppingEventPayloadFilter, Nullable`1 rundown) in /_/src/Tools/dotnet-trace/CommandLine/Commands/CollectCommand.cs:line 286


I'll try again tomorrow

**Comment by jlinkerlathem** [comment]
@jonathanpeppers Ok, I am back on this.  I had tried this before and did not have the same result as I am not.  But, I removed the Mr. Gestures, and we also were using a Sync Fusion Content View, I have replace it with a native content view.  When I run in debug mode it is still taking about 9 seconds to transition.  But, if I run it in release mode it drops down to 3 seconds.  Still too long, but much improved. 

What I don't understand is why Maui has such a huge disparity between debug performance and release performance?  Xamarin did not have this large of a disparity.  I have tried all of the tricks (I think) to get debugging to be more like release.  But, I cannot accurately test my app if the experience is so different between the two.

**Comment by jlinkerlathem** [comment]
For anyone else who is struggling to get debug to work the same as release, all of the posts I have seen about this suggest to add this to your debug build configurations...

<_MauiForceXamlCForDebug>true</_MauiForceXamlCForDebug>

If that does not work for you, then you may find, like I did just now, that it is actually case sensitive, and if you do this instead...

<_MauiForceXamlCForDebug>True</_MauiForceXamlCForDebug>

you may have better luck.  Its moments like these if I was a drinking man I think I would have to stop work and go have a few.  

If you look in the .net Maui source code you will see it checks this value this way...

<_MauiXamlCValidateOnly Condition="'$(Configuration)' == 'Debug' AND '$(_MauiForceXamlCForDebug)' != 'True'">True</_MauiXamlCValidateOnly>

Which is what prompted me to attempt it with a capital T in the true.

This now has me from 9 seconds down to 4 seconds when debugging my page, so maybe this will allow me to better see the real issue.  (Sorry for the multitude of posts.  But I hate when I find a post asking the same question I had, and then the original poster solves their issue but never updates anyone on the process they went through to solve it, or their ultimate resolution.  And I figure someone, somewhere is going to find this post, see the things I tried, and find success with something from it).

**Comment by jonathanpeppers** [comment]
> Which is what prompted me to attempt it with a capital T in the true.

MSBuild is weird, and string comparisons are not case sensitive.

> This now has me from 9 seconds down to 4 seconds when debugging my page,

So, it sounds like "parsed XAML" significantly slows down your app. But this isn't an issue in Release mode at all, just when you are debugging?

**Comment by jlinkerlathem** [comment]
> So, it sounds like "parsed XAML" significantly slows down your app. But this isn't an issue in Release mode at all, just when you are debugging?

Well, it is still slow, but not as slow.  With the True, debugging now behaves the same as Release.  Instead of 9 seconds to transition to the page, it is only about 3 or 4 seconds.  But, it i enable the .net Trace, it slows it back down to 9 seconds and the speed scope file looks about the same.  So, there is still something somewhere slowing it down, but at least now debug and release are giving the same speeds (when not trying to do a trace)

**Comment by jlinkerlathem** [comment]
> [@jlinkerlathem](https://github.com/jlinkerlathem) if I remember correctly `<_MauiForceXamlCForDebug>True</_MauiForceXamlCForDebug>` will break XAML hot reload so keep that in mind.

@OvrBtn Hot Reload has never really worked in our app for some reason, so I have usually kept it turned off.  Every now and then I will get it to work for some extremely basic things, but then it will stop working again, for the most part it has been useless to us.

**Comment by jlinkerlathem** [comment]
To keep anyone watching up to date on my "progress".  Switching my custom content button from xaml to generated in code seemed to help some.  When trace is turned off it will navigate to the page I have mainly been focusing on in about 4 seconds.  When I add code to time various aspects it now find I am generating the UI in approximately 700 ms, but it still takes about 1.5 to 2 seconds to create the view and pass it to Page.Navigation.PushAsync.  But, PushAsync takes about 2 seconds to execute (which is all code in .net Maui, so I cannot really explore why it is taking so long, i looked through the Maui repository and did not see anything right off that would cause this slowness).  

So, quite frankly I am still stuck.  We are converting an existing app from Xamarin to Maui, so all of this is something in Maui, or with the way we ported it, as the same UI and navigation pattern is instantaneous in the Xamarin app.  So, nothing new except still stuck, but wanted to provide an update for anyone following along.

**Comment by jlinkerlathem** [comment]
Spent more time playing around with it tonight.  Looking through the Maui code is difficult to follow exactly what is happening.  But, I am seeing that the page I am navigating away from, there is about 2 seconds or so between the end of OnNavigatingFrom and the beginning of OnNavigatedFrom...  Looking in the Maui code there doesn't seem to be too much happening between those two.  If I am in the correct place, NavigationPage.cs, OnPushAsync calls return Owner.SendHandlerUpdateAsync with a series of calls,

Owner.PushPage(root), 

Owner.SendNavigating(previous page (fires OnNavigatingFrom of the original page)),
Owner.FireDisappearing(previous page (fires OnDisappearing of the original page)),
Owner.FireAppearing(root (fires OnAppearing of the navigating to page)),

Owner.SendNavigated(previousPage, NavigationType.Push (fires OnNavigatedFrom on the original page, then OnNavigatedTo of the new page)),
and lastly Owner?.Pushed?.Invoke(Owner, new NavigationEventArgs(root)); (which fires Pushed on the Application.Current?.MainPage)

If I put an event listener on each of those events they fire in this order with these time stamps...

Minutes:Seconds.Milliseconds

4:0.983 | OriginalPage | OnDisappearing | Start
4:1.8 | OriginalPage | OnDisappearing | End
4:1.16 | Newpage | OnAppearing | Start
4:1.62 | NewPage | OnAppearing | End
4:1.74 | OriginalPage | OnNavigatingFrom | Start
4:1.81 | OriginalPage | OnNavigatingFrom | End

4:3.529 | OriginalPage | OnNavigatedFrom | Start
4:3.536 | OriginalPage | OnNavigatedFrom | End
4:3.543 | NewPage | OnNavigatedTo | Start
4:3.548 | NewPage | OnNavigatedTo | End
4:3.558 | Pushed | Start
4:3.567 | Pushed | End

In the Maui Code, those events are fired off in a method called SendHandlerUpdateAsync, it fires off the first sets of events (processStackChanges, firePostNavigatingEvents), waits for all of those events to have fired, then releases a SemaphoreSlim and then fires fireNavigatedEvents.  All that happens between the events being invoked, and the semaphore being released is 

// Create the request for the handler
var request = new NavigationRequest(immutableNavigationStack, animated);
((IStackNavigation)this).RequestNavigation(request);

// Wait for the handler to finish processing the navigation
// This task completes once the handler calls INavigationView.Finished
await currentNavRequestTaskSource.Task;

So, I am not sure where else to look as to why there is such a delay here...

**Comment by matteocollina** [comment]
I'm using the simulator "Medium phone" with API 35 (Google Play) and It's incredibly faster.

**Comment by jlinkerlathem** [comment]
@matteocollina Is it a simple test app, or an enterprise level app with DI, multiple screens, shareable code, and more than 5 controls on the screen?  If it is a bigger app, do you have a repository you can share?  I am wondering if it is a configuration issue or something small somewhere we missed in our initial set up.  I have tried about 5,000 different things with very little change in speeds.

**Comment by jlinkerlathem** [comment]
@matteocollina Device info doesn't help me here.  I have tested it on actual device, debug and release with same results.   On both a high end Samsung phone, and a low end custom device running Android 7.1.  Same results on both.  Ultimately, its for an application on custom hardware.  It was done in Xamarin and we have to continue to support it and add features, so due to the drop of support of Xamarin we are converting to Maui.

It does seem I have a two fold problem.  Every navigation I make the navigation portion takes 2 second regardless of what I am navigating to.  I have one page I am primarily using as my test page.  It normally has 12 buttons on it (its a keypad for a PIN entry screen).  If I comment out the keypad so it is loading an empty screen, two seconds to navigate.  If the buttons are there it takes 4 seconds total.  This lines up with metrics I gathered from using Debug.WriteLine's with time stamps.  InitializeComponent on the UI takes about 2 seconds when the buttons are there, and then there are about 2 seconds between OnDisappearing and OnDisappeared of the page that is being navigated from.  I have tried looking through the Maui code to see what happens between those two methods, and it appears that may be when it hands it off to Android to replace the first view with the second.

As mentioned before, this app was originally written in Xamarin and has none of these problems.  Same UI, everything.  It is an enterprise level application with base views, View Models, etc.  I have spent about a month on this and made little to no progress.  When googling I find where others are having the same problem with no resolutions.  Some, have an issue with Release being fast and Debug being slow, so they enable _MauiForceXamlCForDebug and fix their issue.  Some, its slow when they navigate to the page the first time and it constructs the view, but subsequent navigations are fast.  But mine is neither of these, it is the same speed in Release and Debug, it is the same speed on the first navigation vs. the 2nd, 3rd, 4th and so on.  Because this is for a custom android device I am only running it on Android, not iOS.  But, the others who seem to have my same issue seem to also not receive any answers.

I am curious if it is some kind of configuration issue, but I cannot find anything vastly different between mine and other samples I see.  Created speed scope files which seemed to back up the idea that it is something with rendering but again, nothing definitive.

**Comment by matteocollina** [comment]
@jlinkerlathem we also made the migration of the app from Xamarin.Forms to .NET Maui but not internally, we used an external agency to do that but the quality of the output was very bad and we spent a lot of hours to investigate in performances. We've found [one library](https://github.com/mono/SkiaSharp.Extended) that worsened the performance so we substituted lottie files with GIF images.
If you want to share the codebase I could make a check.

**Comment by jlinkerlathem** [comment]
@matteocollina We aren't using Skia Sharp on this app, but were planning to use it in another we are migrating, so we will have to keep that in mind.

**Comment by irshadka** [comment]
> [@jonathanpeppers](https://github.com/jonathanpeppers) [@matteocollina](https://github.com/matteocollina) I think I am at the point where I am going to have throw in the towel and see if Microsoft has paid Maui experts support I can hire to look at my code and see what can be fixed. I stripped it down where I did not have any sub content, just flat page, single grid. Removed most of the controls on the page so it was simple as could be. It still took 1 second to load. So, with nothing on the page hardly I am starting at 1 second. Adding the controls I need to the page increases it to 4 seconds. The Xamarin version of the same app, loaded the same UI in 300 ms. _Insert Crying Emoji_

May I know which .net version you are using?

**Comment by jlinkerlathem** [comment]
@irshadka .Net 8, I did try updating to .Net 9 just to verify there was no fix there that would help, but I had the same results.

**Comment by irshadka** [comment]
> [@irshadka](https://github.com/irshadka) .Net 8, I did try updating to .Net 9 just to verify there was no fix there that would help, but I had the same results. (Just checked, and was on Visual Studio 17.12.1, noticed another update is out, so updating to 17.12.2 to make sure I am on the latest)

Can you please go to my profile, I have created a project for learning and testing purposes named MoviezApp. Can you please check if the above-mentioned issue is there in that project in release mode?

**Comment by jlinkerlathem** [comment]
@irshadka I am interested in the totality of the optimizations you have gone through to reach the point of happiness.  

When I run your project it still takes a couple of seconds the first time it loads the search page (almost 1 second for initialize component to fire, and then a half a second between the end of OnNavigatingFrom and OnNavigatedFrom firing on the Homepage.  

Granted, subsequent navigations are quicker, running about 1 second.  But still not as instantaneous as my original Xamarin app.

**Comment by irshadka** [comment]
> [@irshadka](https://github.com/irshadka) I am interested in the totality of the optimizations you have gone through to reach the point of happiness.
> 
> When I run your project it still takes a couple of seconds the first time it loads the search page (almost 1 second for initialize component to fire, and then a half a second between the end of OnNavigatingFrom and OnNavigatedFrom firing on the Homepage.
> 
> Granted, subsequent navigations are quicker, running about 1 second. But still not as instantaneous as my original Xamarin app.

Thanks for the feedback. In my case, our Xamarin project was coded badly with messy constructors and many nested controls in XAML(like nested frames for rounded corners, some libraries for shadows, and many nested layouts). when we migrated as like that to MAUI then it was too slow on Android devices in both debug and release modes.But it was working perfectly on iOS devices and Android Emulators. We cleaned the constructor codes by using singletons and dependency injections and removed the nested controls, keeping only needed controls in UI, replaced the frame with Border, and replaced navigation with Shell navigation, etc. By doing this we got much better navigation speed in Release mode and it made the customer happy.):

**Comment by jlinkerlathem** [comment]
@irshadka And thank you for that information!  We were already using Dependency Injection, but due to base classes and stuff quite a few dependencies are injected into each ViewModel.  So our constructors do have a lot, but its all injected stuff.  We had switched from Shell Navigation to manual navigation due to some issue we encountered with shell (I will have to talk to my other developer to remind myself as to why we did that).  It does sound like we may be in the same boat as far as the controls.  My UI guy likes over engineering controls to handle every possible scenario that can happen.  Though we aren't using Frames, our main button control is made to be able to handle with or without images, and with or without borders.  And it uses two third party libraries.

I still wonder why Maui handles this so much worse than Xamarin.  The UI and its complexity is the same as it was on the Xamarin app.  It was advertised as more or less an in place upgrade with namespace adjustments, but this has not been our experience.  I have been trying to get the Maui source code to build so I could try to step through it and see where the slow down happens and see if there was anything obvious causing it for Android, but have not had any luck there yet.

**Comment by jlinkerlathem** [comment]
@irshadka And for those following along.  We created a simple sample app for testing, to rule out our DI, Bindings, etc.  Loads a screen with a single button, when clicked navigates to a screen with and entry box and 12 buttons (a number pad, 0-9, Delete button, Check mark).  When using Page Navigation and having the buttons as reusable controls, it takes 4 seconds to navigate to it.

If switch to Shell Navigation but keep the reusable controls, it dropped by a second or 2.  But was still 2 to 3 seconds.

If stayed with Page Navigation but flattened the UI to not use the reusable controls, stayed at about 4 seconds.

But, if we flattened the UI and switched to shell navigation, it indeed dropped to a second or less.  Still not ideal or as quick as Xamarin, but we may can work with that.  I am converting our large project back to Shell Navigation, and flattening the UI to see if it works in it as well.

This project is currently only on Android.  Since @irshadka says there's was on iOS as well, but the issue only happens on Android Devices, I am curious why that would be the only place there is an issue.  If it was Android Emulators and Devices, then I would think one problem is it would sound like the reusable controls are not being integrated (flattened) into the xaml before it is passed to the platform renderer.  It would sound like the platform code was handling generating the final xaml that has to be converted.  But it working on the emulator throws a wrench in that idea.

@OvrBtn mentioned backlogged issues, I may try to find them, see which ones seem similar to my own, so I can keep an eye on them and see when fixes are released.

**Comment by irshadka** [comment]
@jlinkerlathem I have observed a small change in navigation speed while disabling the navigation animation. Can you please try it and let me know if this helps?

**Comment by KevThompson2020** [comment]
> [@jlinkerlathem](https://github.com/jlinkerlathem) there is whole thread [#22793](https://github.com/dotnet/maui/discussions/22793) some issues are linked there for example I've mentioned issues where DatePicker is slowing down first navigation and CollectionView generally is slowing down navigation by a lot.
> 
> There is also somewhere an issue opened by MAUI team mentioning mappings being done multiple times and it is still and issue for example Border BackgroundColor property is mapped 8 times on render instead of 1. I'm not sure how big is it's impact though.

That mapping point is quite interesting.  I had a Bug on iOS where a label inside a border would not show in some cases.   Usually lower on the screen.  Could be related to that.

**Comment by jlinkerlathem** [comment]
@irshadka Disabling animations was one of the first things we did.  I don't remember it having a major affect on it.   I just checked in our sample app and it has been with animation turned off.

@KevThompson2020 Could be

**Comment by taublast** [comment]
Need to add my 2 cents here.. When porting xamarin to MAUI that problem of slowness compared to Xamarin is apparent. IMHO the "on iOS it's fine" comes from the fact Apple devices are just on steroids compared to Android. Would bet the problem doesn't come from Android. Maybe handlers with mappers didn't beat renderers.

All the xam->maui estimated "porting time" is multiplied by 2 for me. Only when I replace huge packs of native controls with drawn ones then app starts to "breathe" again.

**Comment by jlinker7** [comment]
This has taken so much for me to get where it is almost kind of ok, that I had to put it on the back burner to do another release with Xamarin to meet some time sensitive demands before I can circle back and attempt this.  The current point I am at is we have completely stripped XAML from the app.  Removed all custom controls.  Flattened the UI out completely.  Removed third party controls.  And yet just creating a default MAUI button takes approximately 50 ms, which doesn't seem bad unless you have 12 of them on the screen and then suddenly you are back up to half a second building the UI and another second and a half to navigate to the page for some reason.  Will be circling back to this in a month or so hopefully.

**Comment by albyrock87** [comment]
If you're using shadows all around, take a look at this: https://github.com/dotnet/maui/pull/26789

**Comment by sandeepnarware** [comment]
We are also facing the same issue, XF app used to work fine, due to EoL we migrated the app to MAUI, but the app is very also on Android, it works fine on iOS devices.  I wonder if its due to the code or complex UI why it work better on iOS its almost 6-10 times slower on the android.

**Comment by jlinker7** [comment]
> We are also facing the same issue, XF app used to work fine, due to EoL we migrated the app to MAUI, but the app is very also on Android, it works fine on iOS devices. I wonder if its due to the code or complex UI why it work better on iOS its almost 6-10 times slower on the android.

Yep.  What I still can't understand is why XAML is suddenly bad.  I looked through the Maui source code and did not see anything that had changed in the XAML Parser, so I cannot wrap my head around why switching from Xamarin to Maui made such a huge difference.  All of the answers I have gotten have been around my XAML not being optimized, or my screens having too much on them, yet, it worked fine in Xamarin.

For those that have followed this post, I previously mentioned stripping XAML out completely and seeing some improvement, but still not to the levels it was with Xamarin.  But, it was taking so long we had to circle bag and do a bug fix release of the Xamarin version to make some customers happy.  Then I will get to switch back to trying to figure this nightmare out.

**Comment by jlinker7** [comment]
To update everyone on my progress, we started converting out mobile app, creating the UI with code instead of Xaml.  On a Samsung Galaxy S22 it runs fine.  But as soon as we put it on a mid to lower end phone it starts crawling.  And looking at the metrics of who uses our app, mid grade phones are our biggest group.  Also, the custom hardware we use for our other app is lower end to keep costs down.

We have opened a support ticket with Microsoft to have them look at our app to tell us if there is something that can be done, and while that is going on I will be evaluating other platforms and making a decision on which direction we will pivot to if support can not get us back on track.

My biggest concern and question mark is still how Maui was reported to be a drop in replacement where the name spaces were the main change, and yet something changed to render UI rendering on anything except a flagship device unusable.

I wish someone on the Maui Team could explain what changed under the hood to make such a huge difference between Xamarin and Maui.

**Comment by Kebechet** [comment]
@jlinker7 in your struggle were you considering Avalonia or Uno if that would make a difference ?

**Comment by taublast** [comment]
We need data comparison of how many times your controls redraw on UI thread in Xamarin vs MAUI, same scenarios.
Up to now for me it looks like excessive UI updates due to bindings causing redraws on almost every property update, at least this is what I observe with my code. Like in Xamarin I change 5 props I get 1 update, in MAUI I change 5 props I get 5 redraws. Like after every prop update the MAUI code is releasing thread for UI to update, the Xamarin alternative of `await Task.Delay(1);`.
That would explain all of the above.

**Comment by taublast** [comment]
@albyrock87 that sounds like a breaking change to me. I use UI-thread and background thread when it is appropriate and would not expect a framework to suddenly switch threads for me, breaking developer control.

Not only this would change the execution flow of any existing Xamarin app code, but also introduce an anti-batching pattern. I can defend both points.

Can you please point to the MAUI code that does that?

**Comment by albyrock87** [comment]
@taublast well not really.. I mean, UI changes can only happen in the UI-thread otherwise iOS (for example) will crash your app.
So MAUI is making sure your app won't crash.

https://github.com/dotnet/maui/blob/1a87feaa2cf5ffd24c5a0194a428e4a029a44bda/src/Controls/src/Core/BindingExpression.cs#L711-L714

That said, this auto-dispatching logic appears to not be used everywhere.. it is on `CollectionView.ItemsSource`, but not on `BindableLayout.ItemsSource`, see the discussion happening here: https://github.com/dotnet/maui/pull/24714

A possible improvement could be batching all the changes happened during a background thread.

That said, you can always import `IDispatcher` in your view model where you're doing things in the background and use it to batch them all, but I get that this is not the same given this way you're also delaying changes at view model level and this also means that the view model has (maybe) too much knowledge of UI matters.

**Comment by taublast** [comment]
Thanks for pointing me to this code! When you say "UI changes can only happen in the UI-thread" do we assume for some reason that values from all bindable static properties in MAUI would land directly into native views? 

So on Android every property change checks for `_looper != Looper.MyLooper();` and then goes updating UI when it was not called from UI thread?

All async commands we used in Xamarin, all the code that is launched by the platform on non UI thread would suddenly just become slower? By creating many UI redraws where in Xamarin there was just one and creating more layout recalculations where in Xamarin there was just one?

I must have misunderstood something here, please correct me.

How I personally use bindings: after new values I  calculate/process data and only then I access platform UI on main thread if needed.  And sometimes it's non needed at all. I don't need to block main thread with processing and calculations and wouldn't expect MAUI to force me to use UI-thread just to exchange data between the viewModel and my controls.

**Comment by albyrock87** [comment]
> Thanks for pointing me to this code! When you say "UI changes can only happen in the UI-thread" do we assume for some reason that values from all bindable static properties in MAUI would land directly into native views? 

That's the general assumption given that all views inherit from `BindableObject` base class.

There's no need to use such a complex `INotifyPropertyChanged` implementation on view model where you can simply use `ObservableObject` from community toolkit.

> So on Android every property change checks for `_looper != Looper.MyLooper();` and then goes updating UI when it was not called from UI thread?

Well, not every property change.

Only the `Binding` mechanism who checks whether you're on the main thread or not and eventually dispatches the `BindableProperty` value set on the next loop.

> All async commands we used in Xamarin, all the code that is launched by the platform on non UI thread would suddenly just become slower? By creating many UI redraws where in Xamarin there was just one and creating more layout recalculations where in Xamarin there was just one?

Afaik, the async code is awaiting by running the continuation on the same async context which should match the UIThread.

I've honestly never verified this with a proper UITest myself, but that's the way it should work 🙂 

> I must have misunderstood something here, please correct me.

That means that as long as you're awaiting starting from a UI event, all of your code should already be running in the UI thread, so that binding dispatching mechanic won't trigger.

> How I personally use bindings: after new values I  calculate/process data and only then I access platform UI on main thread if needed.  And sometimes it's non needed at all. I don't need to block main thread with processing and calculations and wouldn't expect MAUI to force me to use UI-thread just to exchange data between the viewModel and my controls.

MAUI is not forcing you, but it's doing it for you in case you forget to do it when acting from a bg thread.

**Comment by taublast** [comment]
Okay, so to allow people to write thread-agnostic code we are now switching code from background to UI-thread, once set a bindable property from behind?

What I know is that on Xamarin I could set 10 props from behind async command code and only when the context would change, their values would be applied.

And when I wanted a value to be applied instantly I was doing

```
IsBusy = true;
await Task.Delay(1);
```

and I was still processing my API data, mapping it from DTO to viewmodels on background thread. And only after that I would use a nice `MainThread.BeginInvokeOnMainThread` to set the content of an observable collection to the final heavy-processed result.

Nowdays this code is broken? When I set `IsBusy = true;` all mapping DTO -> viewmodels continues on UI-thread making my app freeze?

**Comment by albyrock87** [comment]
@taublast I think at this point a code sample will clarify what I'm saying.

```xml
<VerticalStackLayout>
  <Label Text="{Binding Text}" FontSize="{Binding Size}" />
  <Button Clicked="OnButtonClick" Text="Click me" />
</VerticalStackLayout>
```

```csharp
private async void OnButtonClick(object sender, EventArgs e)
{
	await Task.Delay(100);
	_vm.Size++;
	_vm.Text = $"Hello {_vm.Size}";
}
```

With this code, all properties will be assigned in block, without going through dispatcher.

**Comment by albyrock87** [comment]
@taublast regarding shadows, SR4 contains my fix where if you use a solid background on the node having the shadow the speed will increase *a lot*.

That was the main bottle neck for us on Android.

That said, there are a few PRs opened trying to improve the performance, and what would really help is someone like you to bring real numbers to the table.

Like, if you could show a video comparison between Xamarin and MAUI (if you have the same code behind), or if you're available to try out a few PRs and show whether they'd make your app faster, that'd be great because it'd help me in supporting my changes with the MAUI team.

I don't want to continue spamming this thread, if you're available, please contact me on Discord `albyrock87` and we can chat.

**Comment by jlinker7** [comment]
> [@jlinker7](https://github.com/jlinker7) in your struggle were you considering Avalonia or Uno if that would make a difference ?

@Kebechet  Yes, both of those are on my list to evaluate over the next couple of weeks

**Comment by jlinker7** [comment]
> We need data comparison of how many times your controls redraw on UI thread in Xamarin vs MAUI, same scenarios. Up to now for me it looks like excessive UI updates due to bindings causing redraws on almost every property update, at least this is what I observe with my code. Like in Xamarin I change 5 props I get 1 update, in MAUI I change 5 props I get 5 redraws. Like after every prop update the MAUI code is releasing thread for UI to update, the Xamarin alternative of `await Task.Delay(1);`. That would explain all of the above.

@taublast What's the best way to check the number of redraws?  Is it at the individual control level, or is there a way to check any time the UI is updated?  I would love to check to see if excessive re-draws is a part of the issue.

**Comment by taublast** [comment]
@jlinker7 Im not toally sure today of the best way to do it, but a quick guess would be maybe (let's think together) to create a tracking custom renderer/handler for some our most used control, and override it's both native OnDraw and Invalidate methods of the native Android.Views.View. I would get myself to it by the next week.

**Comment by taublast** [comment]
The main thing that bothers me now is that there is a possibility that that much of the unchanged code that used to work on background thread in Xamarin without blocking the UI, would now executes on UI tread in MAUI. You would not see that in the OnDraw, just the UI would freeze and the app would be slower, since everything would just run on 1 thread where it was running on 2 threads with Xamarin. I guess here we could record the phisical device GPU debug overlay info.. Anyway have to get back to this to investigate more as @albyrock87 pointed out many interesting things that need deeper understanding than one i have today.

**Comment by mnxamdev** [comment]
So is this why we're getting all these "ANR rate exceeds overall bad behavior threshold" in Google Play Console? iOS seems to have similar hanging increase since we release our updated app on MAUI. So what's the fix?

**Comment by jonathanantoine** [comment]
> The main thing that bothers me now is that there is a possibility that that much of the unchanged code that used to work on background thread in Xamarin without blocking the UI, would now executes on UI tread in MAUI. You would not see that in the OnDraw, just the UI would freeze and the app would be slower, since everything would just run on 1 thread where it was running on 2 threads with Xamarin. I guess here we could record the phisical device GPU debug overlay info.. Anyway have to get back to this to investigate more as @albyrock87 pointed out many interesting things that need deeper understanding than one i have today.

I don't understand why you say that it will run on only 1 thread and that what you were doing on 2 threads will now be done in 1 only. 

As I read the code, only the assignment of the property value is done, later, on the main dispatcher (looper on Android) but your code still runs on the same original thread (your background one).


Every time you change a **binded** value, then the UI is updated. It was the case already for Xamarin. Or did I miss something?

**Comment by taublast** [comment]
@jonathanantoine "Every time you change a binded value, then the UI is updated" No, it was not the case for Xamarin when we were running in the background, what was a very common scenario. The background thread execution was not only produced by the developer himself but by the platform operating system in many cases, sometimes creating a difference in execution not only between platforms but between physical devices as well, creating those "this runs on this Android device fine and crashes on this one for some reason".

I think that I misinterpreted what Alberto was saying with "the async code is awaiting by running the continuation on the same async context which should match the UIThread.", and he meant that the remaining code is still running in the background after the value update switches to UI thread. 

Still left with the initial concern of multiple UI updates from the background thread, while in Xamarin we had just one "batch update" by default when exiting the background thread and switching to UI one. Those batch updates were very common in Xamarin so the sudden removal could not leave the usual established performance unaffected.

**Comment by jonathanantoine** [comment]
> @jonathanantoine "Every time you change a binded value, then the UI is updated" No, it was not the case for Xamarin when we were running in the background, what was a very common scenario. The background thread execution was not only produced by the developer himself but by the platform operating system in many cases, sometimes creating a difference in execution not only between platforms but between physical devices as well, creating those "this runs on this Android device fine and crashes on this one for some reason".
> 
> I think that I misinterpreted what Alberto was saying with "the async code is awaiting by running the continuation on the same async context which should match the UIThread.", and he meant that the remaining code is still running in the background after the value update switches to UI thread. 
> 
> Still left with the initial concern of multiple UI updates from the background thread, while in Xamarin we had just one "batch update" by default when exiting the background thread and switching to UI one. Those batch updates were very common in Xamarin so the sudden removal could not leave the usual established performance unaffected.
> 
> 

This is the first time that I heard of this "batching" of changes. Do you have documentation of this ? 

I did a lot of background work in my apps to reduce the load of the UI thread but I had to be very careful to not update a binded property from the background thread to avoid a crash. This is what surprises me.

**Comment by taublast** [comment]
Not a all, this is a behaivoural observation. A lot of aspects in how binding worked, let's say work, as I still continue use Xamarin with my clients who do, might look surprising. Issues like "my bindings are not working" etc. 
There are properties that you should never set from a background thread and i think you might relate to them, when saying about being careful, like accessing those directly accessing UI views and observable collections. But in Xamarin you could set 10 props that affect just the UI logic have them affect your UI on next redraw, when you enter the UI thread. I call this a "batch update", totally un-designed, undocumented, this is just how it works and it's fast. And when you start to force-assign absolutely all props on UI thread you get a performance degradation.

**Comment by jlinker7** [comment]
I am still doing some research to figure out precisely what's going on in mine.  But I found OnDraw for my controls was only firing once, and the timing was not bad.  But, I noticed the first OnDraw was firing way after OnNavigatedTo.  So, I put debug writelines in all events on my page.  I found that OnPropertyChanged fires 12 times.  But then, OnChildMeasureInvalidated is firing 55 times for the grid on the screen (still working to validate which grid on the screen as there are more than one).  And this appears to be where all of the time goes in my code.  Still working on narrowing it down.

**Comment by taublast** [comment]
Can anyone please comment if the following method is viable:

```csharp
public class VStackTrace : StackLayout
{
    private long invalidated;

    public VStackTrace()
    {
        this.MeasureInvalidated += (s, a) =>
        {
            Super.Log($"[VStackTrace] {this.GetType().Name} MeasureInvalidated {++invalidated}");
        };
    }
}
```

I subclassed the control that holds page content.

In Xamarin the last i I got for this type was `[0:] [VStackTrace] ContentDevelop MeasureInvalidated 8`

In MAUI:

![Image](https://github.com/user-attachments/assets/49187464-d1a0-4b2a-bc18-49dea60795fd)

**Comment by albyrock87** [comment]
@taublast it's not.. leave alone MeasureInvalidated as it has basically no meaning in MAUI.
It has an important impact on iOS CollectionView but soon I'll bring a PR which will make it irrelevant.

There are three things which can be measured:
- measure pass
- arrange pass
- draw pass

Regarding the arrange pass, you can check what I've done in this UI test:
https://github.com/dotnet/maui/blob/main/src/Controls/tests/TestCases.HostApp/Issues/RenderingPerformance.xaml.cs

A similar approach can be used on the measure pass by subclassing an handler and overridding `GetDesiredSize`.
https://github.com/dotnet/maui/blob/2f3bb5f87bcb5587b446e24f4afa7ee0eb6bf412/src/Core/src/Handlers/View/ViewHandlerOfT.iOS.cs#L19

Regarding draw, that is platform dependant.

**Comment by taublast** [comment]
@BretJohnson maybe a compromise could be in place let create custom controls that do not require UI thread.  
A nice way could be to create context-aware binding extensions, to apply all "stay on UI thread" logic to `VisualElement`. All controls deriving from it would benefit from the WPF approach and `BindableObject` alone would be exempt of assumptions and restrictions, and have most of it's logic unsealed. This would make MAUI an open framework, and we could build anything with it.   
For instance, since Maui.Core references Graphics we might make full use of it: drawn controls do not need UI thread and can consume bindings in background.

**Comment by PureWeen** [comment]
I'm going to close this issue for now since there's not a specific issue. 

I created a tracking issue here https://github.com/dotnet/maui/issues/28091. 
We've been having some discussions around ways to add some information as .NET MAUI components are rendered to give better information to users at runtime. 

If you're still having a specific issue and you can recreate a new issue and repro that would be very helpful.

**Comment by BretJohnson** [comment]
@PureWeen - what do you think of my suggestion here https://github.com/dotnet/maui/issues/18838#issuecomment-2687930931 ? Is a warning appropriate there.?

**Comment by PureWeen** [comment]
> [@PureWeen](https://github.com/PureWeen) - what do you think of my suggestion here [#18838 (comment)](https://github.com/dotnet/maui/issues/18838#issuecomment-2687930931) ? Is a warning appropriate there, when calling from a non-UI thread and the thread switch needs to happen? I wasn't sure of the intent (as it's different than WPF which just requires everything to be on the UI thread). In any case, outputting some kind of message as part of [#28091](https://github.com/dotnet/maui/issues/28091) seems appropriate here.

I've referenced your comment in the issue I created.

**Comment by jlinker7** [comment]
For any one following my progress.  In working with Microsoft Support to try to figure this our, I created am emulator that matched our physical device.  I created a minimum reproducible sample.

When i ran it at first it was super fast.  And I was confused.  Then I realized, the emulator I created I had accidentally used x86 for the processor instead of armabi-v7a. When i switched it to armabi-v7a, it immediately slows to a crawl.  I took our original code base that is the issue, ran it on the x86 emulator, runs fine, ran it on the arm and it slows down.

Unfortunately, the device we have to run it on, runs armabi-v7a.  But, if this gives insight to anyone on what may be going on, the processor seems to cause the app to behave completely different speed wise.   

I had a few different scenarios I was testing, and the results ranged from 6 to 13 seconds on arm, to 500 milliseconds to 1.5 seconds on x86.  The processor being the only change.

**Comment by jlinker7** [comment]
https://github.com/jlinker7/MauiSpeedTest

For anyone who wants to play with my sample app, here is a link to a repository.  It has a shell version and a page navigation version.  When ran there are two buttons, one for loading a page with 12 buttons using custom controls, and one to load it with a flat ui.

If you create an emulator that is ARM and Android 7.1 the speed issue is present.  If you run it on an emulator that is x86 you will see if is quick.

**Comment by jgold6** [comment]
> If you create an emulator that is ARM and Android 7.1 the speed issue is present. If you run it on an emulator that is x86 you will see if is quick.

Is this on the same computer host? 
An ARM emulator will run very very slowly on an x86/64 host (e.g. Windows desktop PC)
Likewise, an x86/64 emulator will run very very slowly on an ARM device (e.g. Mac with Apple silicon)

**Comment by bkaankose** [comment]
> IMO it'd be best if MAUI behaved like this here:
> 
> * When calling from a non-UI, it's OK for MAUI to switch to a UI thread. That's (arguably) better than crashing or producing incorrect results. WPF, for instance, doesn't do this and can crash or produce incorrect results
> * BUT when MAUI does this it should output a Debug message, maybe flagged as a warning, to help warn the developer that a thread switch occurred because the API was called from the "wrong" thread. That can help indicate to devs that encounter this that it'd be better for them to switch to the UI thread in the first place, batching calls when possible so that thread switch only happens once. The warning message helps serve as a clue to possible performance problems. And as WPF (and Xamarin) devs aren't expecting an automatic thread switch (since those platforms don't do that), the warning message also helps indicate that MAUI is different in how it behaves.

I would love MAUI to crash when devs did something they are not supposed to do, and don't automatically switch threads at all, but this will be a super breaking change at this point. All other xaml based frameworks do it. You call INotifyPropertyChanged or INotifyCollectionChanged on a background thread? Game is over. Go and fix it.

The more responsibility MAUI takes on behalf of developers the more it is getting unpredictable and bloated. It's harder to diagnose, you don't really know what is causing the issues, and you have less control over it.

Maybe a flag (off by default) in MSBuild is a better option for making those calls crash.
